core/src/main/java/org/apache/iceberg/BaseFilesTable.java

```diff
@@ -54,7 +56,23 @@ public Schema schema() {
       schema = TypeUtil.selectNot(schema, Sets.newHashSet(DataFile.PARTITION_ID));
     }
 
-    return TypeUtil.join(schema, MetricsUtil.readableMetricsSchema(table().schema(), schema));
+    return withDerivedColumns(schema);
+  }
+
+  private Schema withDerivedColumns(Schema schema) {
+    AtomicInteger nextId = new AtomicInteger(schema.highestFieldId());
```

Comment:
I think we dont need the complexity of nextId and can make a constant field id (and even make the whole NestedField into a constant to reduce code in this method).

Reply:
Ended up using a method which take highestFieldId for fieldId